### ISSUE TITLE
Add event history to Rust runtime interface

### ISSUE THREAD

**Issue by walkie** [frame-lang/frame_transpiler#62]
For visualization and debugging, it could be useful to have access to the history of Frame events for a given state machine.

This would be a bit more work than it seems because currently events live on the stack in the Rust backend, so we'd need to change the generated code to use `Rc` everywhere for events when the runtime system is enabled.

**Comment by frame-lang** [comment]
Interesting to explore, but perhaps this might be better served by a standardized logging mechanism/abstraction? It is one of the use cases I think a language supported communications scheme/abstraction might support well.

**Comment by walkie** [comment]
Turns out I'm sort of deep into implementing this already! 😳 

https://github.com/frame-lang/frame_transpiler/tree/event-history

We needed a way to access the last transition per: #61 

I could (and maybe should) have added this in a really lightweight way by just tracking the ID of the last transition, but then we wouldn't have access to things like enter/exit args. So, I started adding the ability to save the whole transition event... and from there it made sense to provide a runtime interface for all Frame events... and then it was just a short jump to being able to save those as well.

It turns out that all of this was even more work than I guessed when I said it would be more work than it seems... but I think it opens up a lot of possibilities besides just event history. In particular, it should be easy from here to add a callback notification system for arbitrary Frame events.

Anyway, I think your point is valid that a standardized logging mechanism might be a better way to handle event tracing, and maybe we still want that. But I think there are other benefits to having a runtime interface to events, and event history is cheap to add once we have that.

The way I've implemented it is completely configurable, and I'll set the default config to retain no event history and only the immediately previous transition.

I've only implemented all of this in `frame_runtime` so far and haven't updated the Rust backend to actually use it yet.
